daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -210,25 +211,55 @@ private[lf] object Speedy {
       }
     }
 
+    // markBase is called when pushing a continuation which requires access to temporaries
+    // currently on the env-stack.  After this call, envBase is set to the current
+    // env.size. The old envBase is returned so it can be restored later by the caller.
     @inline
-    def restoreEnv(
-        frameToBeRestored: Frame,
-        actualsToBeRestored: Actuals,
-        envSizeToBeRestored: Int,
-    ): Unit = {
-      // Restore the frame and actuals to there state when the continuation was created.
-      frame = frameToBeRestored
-      actuals = actualsToBeRestored
-      // Pop the env-stack back to the size it was when the continuation was created.
-      if (envSizeToBeRestored != env.size) {
-        val count = env.size - envSizeToBeRestored
-        if (count < 1) {
-          crash(s"restoreEnv, unexpected negative count: $count!")
-        }
+    def markBase(): Int = {
+      val oldBase = this.envBase
+      val newBase = this.env.size
+      val increase = newBase - oldBase
+      if (increase < 0) {
+        crash(s"markBase: $oldBase -> $newBase -- NOT AN INCREASE")
+      }
+      this.envBase = newBase
+      oldBase
+    }
+
+    // restoreBase is called when executing a continuation which previously saved the
+    // value of envBase (by calling markBase).
+    @inline
+    def restoreBase(envBase: Int): Unit = {
+      val reduction = this.envBase - envBase
+      if (reduction < 0) {
+        crash(s"restoreBase: ${this.envBase} -> ${envBase} -- NOT A REDUCTION")
+      }
+      this.envBase = envBase
+    }
+
+    // popTempStackToBase is called when we begin a new code-context which does not need
+    // to access any temporaries pushed to the stack by the current code-context. This
+    // occurs either when returning to the top continuation on the kontStack or when
+    // entering (tail-calling) a closure.
+    @inline
+    def popTempStackToBase(): Unit = {
+      val envSizeToBeRestored = this.envBase
+      val count = env.size - envSizeToBeRestored
+      if (count < 0) {
+        crash(s"popTempStackToBase: ${env.size} --> ${envSizeToBeRestored} -- WRONG DIRECTION")
+      }
+      if (count > 0) {
         env.subList(envSizeToBeRestored, env.size).clear
```

Comment:
```
daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:252: error: method removeRange in class ArrayList cannot be accessed in com.daml.lf.speedy.Speedy.Env
 Access to protected method removeRange not permitted because
 enclosing class Machine in object Speedy is not a subclass of
 class ArrayList in package util where target is defined
        env.removeRange(envSizeToBeRestored, env.size)
```

Also, the `subList`/`clear` pattern is used also in `tryHandleException`

If we do pursue this improvement, let's leave it to a future PR.

Reply:
Yeah, let's leave it for a separate PR.